File: technical/reverse_c_style_string/golang/reverse_c_style_string_test.go
package reverse_c_style_string

import "testing"

func equal(a, b []byte, size int) bool {
	for i := 0; i < size; i++ {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func TestReverse(t *testing.T) {
	empty := []byte("")
	empty = append(empty, byte(0)) // this makes it cstyle
	one := []byte("1")
	one = append(one, byte(0))
	even := []byte("even")
	even = append(even, byte(0))
	odd := []byte("odd")
	odd = append(odd, byte(0))
	extraNulls := []byte("extra")
	extraNulls = append(extraNulls, byte(0))
	extraNulls = append(extraNulls, byte(0))

	if !equal(Reverse(empty), []byte(""), 0) {
		t.Error()
	}
	if !equal(Reverse(one), []byte("1"), 1) {
		t.Error()
	}
	if !equal(Reverse(even), []byte("neve"), 4) {
		t.Error()
	}
	if !equal(Reverse(odd), []byte("ddo"), 3) {
		t.Error()
	}
	if !equal(Reverse(extraNulls), []byte("artxe"), 5) {
		t.Error()
	}
}

File: technical/reverse_c_style_string/c++/reverse_c_style_string_test.cpp
#include <stdio.h>
#include <assert.h>
#include <cstring>

void testEmpty();
void testOne();
void testEven();
void testOdd();
void testExtraNulls();

void reverse(char*);

int main() {
  testEmpty();
  testOne();
  testEven();
  testOdd();
  testExtraNulls();
  printf("PASS\n");
  return 0;
}

void testEmpty() {
  char empty[] = "";
  reverse(empty);
  assert(strcmp(empty, "") == 0);
}

void testOne() {
  char one[] = "1";
  reverse(one);
  assert(strcmp(one, "1") == 0);
}

void testEven() {
  char even[] = "even";
  reverse(even);
  assert(strcmp(even, "neve") == 0);
}

void testOdd() {
  char odd[] = "odd";
  reverse(odd);
  assert(strcmp(odd, "ddo") == 0);
}

void testExtraNulls() {
  char foo[] = "hello";
  foo[4] = 0;
  reverse(foo);
  assert(strcmp(foo, "lleh") == 0);
}